src/main/java/org/apache/spark/shuffle/ucx/memory/MemoryPool.java

```diff
@@ -0,0 +1,155 @@
+ * See file LICENSE for terms.
+ */
+package org.apache.spark.shuffle.ucx.memory;
+
+import org.apache.spark.shuffle.UcxShuffleConf;
+import org.apache.spark.unsafe.Platform;
+import org.openucx.jucx.ucp.UcpContext;
+import org.openucx.jucx.ucp.UcpMemory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.nio.ByteBuffer;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicInteger;
+
+
+/**
+ * Utility class to reuse and preallocate registered memory to avoid memory allocation
+ * and registration during shuffle phase.
+ */
+public class MemoryPool implements Closeable {
+  private static final Logger logger = LoggerFactory.getLogger(MemoryPool.class);
+
+  @Override
+  public void close() {
+    for (AllocatorStack stack: allocStackMap.values()) {
+      stack.close();
+      logger.info("Stack of size {}. " +
+          "Total requests: {}, total allocations: {}, preAllocations: {}",
+        stack.length, stack.totalRequests.get(), stack.totalAlloc.get(), stack.preAllocs.get());
+    }
+    allocStackMap.clear();
+  }
+
+  private class AllocatorStack implements Closeable {
+    private final AtomicInteger totalRequests = new AtomicInteger(0);
+    private final AtomicInteger totalAlloc = new AtomicInteger(0);
+    private final AtomicInteger preAllocs = new AtomicInteger(0);
+    private final ConcurrentLinkedDeque<RegisteredMemory> stack = new ConcurrentLinkedDeque<>();
+    private final int length;
+
+    private AllocatorStack(int length) {
+      this.length = length;
+    }
+
+    private RegisteredMemory get() {
+      RegisteredMemory result = stack.pollFirst();
+      if (result == null) {
+        ByteBuffer buffer = Platform.allocateDirectBuffer(length);
+        UcpMemory memory = null;
+        if (conf.preregisterMemory()) {
+          memory = context.registerMemory(buffer);
+        }
+        result = new RegisteredMemory(new AtomicInteger(1), memory, buffer);
+        totalAlloc.incrementAndGet();
+      } else {
+        result.getRefCount().incrementAndGet();
+      }
+      totalRequests.incrementAndGet();
+      return result;
+    }
+
+    private void put(RegisteredMemory registeredMemory) {
+      registeredMemory.getRefCount().decrementAndGet();
+      stack.addLast(registeredMemory);
+    }
+
+    private void preallocate(int numBuffers) {
+      while ((long)length * (long)numBuffers > Integer.MAX_VALUE) {
+        numBuffers--;
+      }
+      ByteBuffer buffer = Platform.allocateDirectBuffer(length * numBuffers);
+      UcpMemory memory = null;
+      if (conf.preregisterMemory()) {
+        memory = context.registerMemory(buffer);
+      }
+      AtomicInteger refCount = new AtomicInteger(numBuffers);
+      for (int i = 0; i < numBuffers; i++) {
+        buffer.position(i * length).limit(i * length + length);
+        final ByteBuffer slice = buffer.slice();
```

Comment:
there is some code duplication between this function and get() method

Reply:
Just 2 lines. There's different logic there. In get it'll allocate and register 1 buffer. Here it'll register contiguous memory chunk and cut between buffers. E.g. 10 buffers for 1Mb: it'll register 10mb memory and split for 10 buffers of 1 mb.

Reply:
can we have 1 common function to create and register a new buffer, used by both get() and preallocate()?

Reply:
Added 2 functions, since RegisteredMemory requires both buffer and memory pointers. The memory could be null if no preregister specified in the configuration.